tnqvm/visitors/exatn-mps/ExaTnMpsVisitor.cpp

```diff
@@ -0,0 +1,1248 @@
+    ket.rename("MPSket");
+    const bool evaledOk = exatn::evaluateSync(ket);
+    assert(evaledOk); 
+    auto talsh_tensor = exatn::getLocalTensor(ket.getTensor(0)->getName()); 
+    if (talsh_tensor)
+    {
+        const std::complex<double>* body_ptr;
+        const bool access_granted = talsh_tensor->getDataAccessHostConst(&body_ptr); 
+        if (!access_granted)
+        {
+            std::cout << "Failed to retrieve tensor data!!!\n";
+        }
+        else
+        {
+            for (int i = 0; i < talsh_tensor->getVolume(); ++i)
+            {
+                const auto& elem = body_ptr[i];
+                std::cout << elem << "\n";
+            }
+        }
+        
+    }
+    else
+    {
+        std::cout << "Failed to retrieve tensor data!!!\n";
+    } 
+}
+
+void ExatnMpsVisitor::finalize() 
+{ 
+    if (m_aggrerateEnabled)
+    {
+        m_aggrerator.flushAll();    
+        evaluateTensorNetwork(*m_tensorNetwork, m_stateVec);
+    }
+
+    if (m_buffer->size() < MAX_NUMBER_QUBITS_FOR_STATE_VEC) 
+    {
+        exatn::TensorNetwork ket(*m_tensorNetwork);
+        ket.rename("MPSket");
+        const bool evaledOk = exatn::evaluateSync(ket);
+        assert(evaledOk); 
+        const auto tensorData = getTensorData(ket.getTensor(0)->getName());
+        // Simulate measurement by full tensor contraction (to get the state vector)
+        // we can also implement repetitive bit count sampling 
+        // (will require many contractions but don't require large memory allocation) 
+        // DEBUG:
+        // printStateVec();
+
+        if (!m_measureQubits.empty())
+        {
+            addMeasureBitStringProbability(m_measureQubits, tensorData, m_shotCount);
+        }
+    }
+    else
+    {
+        if (!m_measureQubits.empty())
+        {
+            std::cout << "Simulating bit string by MPS tensor contraction\n";
+            for (int i = 0; i < m_shotCount; ++i)
+            {
+                const auto convertToBitString = [](const std::vector<uint8_t>& in_bitVec){
+                    std::string result;
+                    for (const auto& bit : in_bitVec)
+                    {
+                        result.append(std::to_string(bit));
+                    }
+                    return result;
+                };
+
+                m_buffer->appendMeasurement(convertToBitString(getMeasureSample(m_measureQubits)));
+            }
+        }
+    }
+    
+
+    for (int i = 0; i < m_buffer->size(); ++i)
+    {
+        const bool qTensorDestroyed = exatn::destroyTensor("Q" + std::to_string(i));
+        assert(qTensorDestroyed);
+    }
+}
+
+void ExatnMpsVisitor::visit(Identity& in_IdentityGate) 
+{ 
+    if (m_aggrerateEnabled)
+    {
+        m_aggrerator.addGate(&in_IdentityGate);
+    }
+    else
+    {
+        // Skip Identity gate
+    }
+}
+
+void ExatnMpsVisitor::visit(Hadamard& in_HadamardGate) 
+{ 
+    if (m_aggrerateEnabled)
+    {
+        m_aggrerator.addGate(&in_HadamardGate);
```

Comment:
m_aggrerator -> m_aggregator

Reply:
I cannot spell this word :) Fixed all the typos in https://github.com/ORNL-QCI/tnqvm/pull/32